cpp/CMakeLists.txt

```diff
@@ -191,8 +191,11 @@ endif(BUILD_TESTS)
 ###################################################################################################
 # - include paths ---------------------------------------------------------------------------------
 
-include_directories("${CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES}"
-                    "${CMAKE_BINARY_DIR}/include"
+if(CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES)
+	include_directories("${CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES}")
+endif()
```

Comment:
I don't like this because if `CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES` is empty it can mean that the system doesn't have a usable CUDA installation as opposed to being installed in standard system locations.

From my perspective it would be better to use https://cmake.org/cmake/help/v3.12/module/CheckLanguage.html to check if CUDA can be enabled and then find the relevant directories to include accordingly.

Reply:
@kkraus14 : I think you may be mistaken, because we enable CUDA  as a language with the argument to the `project()` command. So the relevant check is probably (not 100% sure) performed already.  So I believe it isn't possible for `CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES` to be empty except in "legitimate" cases.

Also note that the link says:

>  If `CMAKE_<lang>_COMPILER` is already defined the check does nothing.

And in our case, the CUDA compiler is determined by CMake, i.e. `check_language()` would do nothing in our case IIANM.

caveat: I'm not a CMake guru, so I might be wrong here.

Reply:
@kkraus14 : So, Mark is convinced :-)

Reply:
@eyalroz the `project()` command does an `enable_language()` call, not a `check_language()` call. And according to the CMake documentation you can use `check_language()` for support before enabling a language.

Reply:
@kkraus14 : Ok, so how about we add:
```
if(not(CMAKE_CUDA_COMPILER))
  message(SEND_ERROR "CMake failed to determine how to compile CUDA")
endif()
```
at the top of our `CMakeLists.txt`'s ?

Reply:
> @kkraus14 : Ok, so how about we add:
> 
> ```
> if(not(CMAKE_CUDA_COMPILER))
>   message(SEND_ERROR "CMake failed to determine how to compile CUDA")
> endif()
> ```
> 
> at the top of our `CMakeLists.txt`'s ?

That works, with the caveat that we should reword it to: `CMake failed to find the CUDA compiler`

Reply:
@kkraus14 : By the way - for some reason, the CMake our environment uses isn't willing to acknowledge a `check_language()` command. But I'll add the three lines.